### PR TITLE
Feature/templates

### PR DESCRIPTION
**Pull request by toroworx**

### CODE REVIEW

**Comment by toroworx** [comment]
Additions to PR https://github.com/akeeba/buildfiles/pull/18

File: phing/common.xml [hunk 1]
```diff
@@ -103,24 +112,70 @@ Akeeba Build Files
...
 	-->
-	<target name="setup-templates" description="Setup build templates" depends="setup-properties">
+	<target name="setup-templates-general" description="Setup general build templates"
+					depends="setup-properties">
 		<copy todir="${dirs.root}" overwrite="true">
 			<fileset dir="${dirs.build.templates}">
 				<include name="**" />
+				<exclude name="**_core.xml" />
+				<exclude name="**_pro.xml" />
 				<exclude name="link.php" />
```
Comment (toroworx) [inline phing/common.xml]: The test was good but I've overlooked this and I think I also have to change this to **/*_core.xml in stead of **_core.xml. Same for the pro.

**Comment by nikosdion** [comment]
I'm afraid it's still wrong.

Once you set up a filterchain you can no longer modify it. That's why we need to put the filterchains inline the transformation in the first place. What you did will have the Pro flag stuck to the same value in both the Core and Pro version.

Moreover you do not take into account the case where I need Pro/Core manifests for some plugins / modules but not all.

I think that the kludges you are introducing outweigh the usefulness of the approach, without any tangible benefit. If you want to add more variables in _your_ use case you can simply create a folder, let's say `build/my_templates`. Add a prerequisite task after setup-properties in the "git" task which applies filters for all variables EXCEPT date/version while copying these files into build/templates.

In fact, you can generalize it. Create a Phing XML which inherits from common.xml and applies your customziations. Your projects can now have _their_ Phing files inherit from that XML file. This way you can reap the benefits of common.xml and your customizations which make sense for your projects without the need for either of us to jump through hoops to release our software.

**Comment by toroworx** [comment]
> I'm afraid it's still wrong.

Uh oh... ;)

> Once you set up a filterchain you can no longer modify it. That's why we need to put the filterchains inline the transformation in the first place.

That's true but you shouldn't need to be able to modify it. The variables like the `PRO` key are still variables which are regenerated for every core/pro/regular version. 

>  What you did will have the Pro flag stuck to the same value in both the Core and Pro version.

Negative. I've tested and confirmed it for component building. This addresses your issue:
`<target name="package-com-builder" depends="setup-templates-corepro" description="Builds a component package archive">`

> Moreover you do not take into account the case where I need Pro/Core manifests for some plugins / modules but not all.

You got me there partly with my pants down. I forgot these. This can be fixed by moving depends from `package-pkg` to `package-pkg-builder` (whoops!). If you add that you got the freedom you'll need. Don't need a Core/Pro manifest? Don't add a `something-core.xml` manifest to your `/build/templates` directory. You can just use `something.xml`. This can be done for *any* subfolder in `templates`. So for any plugin/module/whatever :-D.

Really, this code is really useful :-). Yes you got quite some lines added to the buildfile, but it gives you a lot of freedom. For build.xml of FOF for example, you can remove all these lines:

		
		<copy file="./templates/lib_fof30.xml" tofile="${dirs.root}/lib_fof30.xml" overwrite="true">
			<filterchain>
				<replacetokens begintoken="##" endtoken="##">
					<token key="DATE" value="${build.date}" />
					<token key="VERSION" value="${version}" />
				</replacetokens>
			</filterchain>
		</copy>
		<copy file="${dirs.root}/lib_fof30.xml" tofile="${dirs.library}/lib_fof30.xml" overwrite="true" />

		
		<copy file="./templates/version.txt" tofile="${dirs.library}/version.txt" overwrite="true">
			<filterchain>
				<replacetokens begintoken="##" endtoken="##">
					<token key="DATE" value="${build.date}" />
					<token key="VERSION" value="${version}" />
				</replacetokens>
			</filterchain>
		</copy>

		
		<copy file="./templates/include.php" tofile="${dirs.library}/include.php" overwrite="true">
			<filterchain>
				<replacetokens begintoken="##" endtoken="##">
					<token key="DATE" value="${build.date}" />
					<token key="VERSION" value="${version}" />
				</replacetokens>
			</filterchain>
		</copy>

Plus the fact you won't need a single line for future files :)

I already got a fork of your project and I customized the `build.xml`. However I'd love to recontribute some of my improvements. And yes of course also a little bit self interest, it's easier to keep up with your improvements ;)

**Comment by nikosdion** [comment]
> That's true but you shouldn't need to be able to modify it. The variables like the PRO key are still variables which are regenerated for every core/pro/regular version.

Wrong. Core and Pro are "flavors" of the same version. When I release Akeeba Something 1.2.3 I *MUST*, at the same time, generate BOTH Akeeba Something Core 1.2.3 and Akeeba Something Pro 1.2.3 (along with its documentation in PDF format). Otherwise I cannot do `phing all release -Dversion=1.2.3` and I am back in the Dark Ages when I had to upload stuff manually.

> For build.xml of FOF for example, you can remove all these lines

Actually that was just me being stupid. FOF doesn't have changing flags. I can just declare the filterchain once.

HOWEVER! Where you declare the filterchain DOES matter as it's parsed exactly when it's first declared. This means that you cannot declare a filterchain BEFORE setup-properties runs UNLESS you manually specify BOTH version AND date.

The thing is, that's not always the case with me. In fact, most of the time I want to do `phing ftpdeploy` which builds a dev releases and uploads the individual Core and Pro packages into different folders over SFTP. Having to declare version numbers and dates manually makes common.xml useless for me.

I really do understand what you are doing and what is your use case. However, it's not compatible with how we find it easier to build software here at Akeeba Ltd. 

Refactoring all of our repositories is a daunting task that takes _days_. We're talking about 30 repositories. Each one has to be adjusted and made sure the software is built correctly. This means build something with the old version, switch to the new version of buildfiles, adjust build.xml, build again, recursively extract all ZIP files, recursively compare ZIP files, rinse and repeat.

If you break b/c you are making me do useless busy work which is prone to oversights (with 30 repositories I'd say an average of 2 issues across all of them has a 99% chance of happening according to my experience). An oversight in a repository can cost me thousands of Euros in lost business. So of course I don't want anything that breaks b/c because it _costs so darn much_.

In the end of the day it's _Akeeba_ Build Files so we will err on the side of maintaining b/c and well tested solutions for our software. Feel free to fork or use the intermediate XML file solution I mentioned. I won't include this PR or its replacement because it solves no problem I have and introduces some that I don't want to have.

**Comment by toroworx** [comment]
Thanks for your patience, Nicholas.

> The variables like the PRO key are still variables which are regenerated for every core/pro/regular version.

> Wrong. Core and Pro are "flavors" of the same version. When I release Akeeba Something 1.2.3 I MUST, at the same time, generate BOTH Akeeba Something Core 1.2.3 and Akeeba Something Pro 1.2.3 (along with its documentation in PDF format). Otherwise I cannot do phing all release -Dversion=1.2.3 and I am back in the Dark Ages when I had to upload stuff manually.

I didn't explain it correctly. Only the `token key="PRO" value="${build.is_pro}` variable is changing when building in order to build the core and pro versions. The version number and date stay the same and I confirmed this is working. My output for a core/pro `phing git -Dversion=1.2.3` task looks like (after the fix):

- com_example-core.zip
- com_example-pro.zip
- pkg_example-1.2.3-core.zip
- pkg_example-1.2.3-pro.zip
- ....

XML manifest differs for the Core and Pro zip packages. The different version.php files also works.

> Wrong. Core and Pro are "flavors" of the same version. When I release Akeeba Something 1.2.3 I MUST, at the same time, generate BOTH Akeeba Something Core 1.2.3 and Akeeba Something Pro 1.2.3 (along with its documentation in PDF format). Otherwise I cannot do phing all release -Dversion=1.2.3 and I am back in the Dark Ages when I had to upload stuff manually.

I see yes. This is working for me after the modification written in the next point.

> HOWEVER! Where you declare the filterchain DOES matter as it's parsed exactly when it's first declared. This means that you cannot declare a filterchain BEFORE setup-properties runs UNLESS you manually specify BOTH version AND date.

Mzzz. I see (so that's how *you* declare the version number!), when I build it with your method then it indeed doesn't work. I've fixed it by moving the filterchain, calling it in `setup-properties` and adding `depends="setup-properties"` to `setup-templates-corepro` task. That works now, without the need to declare the date.

> Having to declare version numbers and dates manually makes common.xml useless for me.

I agree, you get tired of that really fast. That's fixed now.

> Refactoring all of our repositories is a daunting task that takes days. We're talking about 30 repositories. Each one has to be adjusted and made sure the software is built correctly. This means build something with the old version, switch to the new version of buildfiles, adjust build.xml, build again, recursively extract all ZIP files, recursively compare ZIP files, rinse and repeat.

> If you break b/c you are making me do useless busy work which is prone to oversights (with 30 repositories I'd say an average of 2 issues across all of them has a 99% chance of happening according to my experience). An oversight in a repository can cost me thousands of Euros in lost business. So of course I don't want anything that breaks b/c because it costs so darn much.

Yeah that's a hell of a work... and it's indeed likely that it won't go without issues. Breaking the b/c regarding the `/build/templates` file structure can't be fixed with this PR. Neither the need to remove code from your `build.xml` files.  I can't see your commercial build files, but as far as I understand the other issues are fixed or can be fixed without too much hustle. But yes adapting all you repositories is quite some work and I understand you rather not do that.

> In the end of the day it's Akeeba Build Files so we will err on the side of maintaining b/c and well tested solutions for our software. Feel free to fork or use the intermediate XML file solution I mentioned. I won't include this PR or its replacement because it solves no problem I have and introduces some that I don't want to have.

No hard feelings. I understand your point of view. I already have a fork anyway. Maybe a good idea for Akeeba Build Tools 2.0 ;)

Cheers!

**Pull request closed without merging**